### DISCUSSION TITLE
WIC support

### DISCUSSION THREAD

**Discussion by windelbouwman** [OE4T/meta-tegra#620] [Ideas]
As mentioned in a ticket, we could start a discussion on how to enable wic support.

As mentioned [here](https://docs.yoctoproject.org/dev-manual/common-tasks.html#creating-partitioned-images-using-wic), WIC seems to be the tool of choice for partitioning a device. Given the fact that the tegra can have an APP and APP_b partition to support A/B updates, it might make sense to implement the logic such that wic can be used to configure this. It would be helpful to yocto users if they could use WIC to set this up.

I see to possible use cases:
- developer wants to setup A/B root FS and a data partition
- developer wants to simply flash entire device (current situration)

It would be great if the partitioning of a tegra device could be somewhat flexible configured, without having to deal with modifying flash.xml files via overrides.

Any ideas / directions on how to do this?

**Comment by madisongh** [comment]
One aspect of the Jetson flashing tools that might be hard to deal with is their interactivity with the device - decisions about which bootloader binaries/configs, device trees, etc. to load onto the device can only be made once the flashing tools have communicated with it and extracted model/revision information from the on-SOM EEPROM. There are ways around that - hard-coding some environment variables before calling on the flashing script, for instance - but you could end up with an unbootable device if there's a mismatch (potentially, at least). I don't know how well wic would work if the content going into the generated image is not known a priori.

It would certainly be useful to have something that makes it easier to adjust partition sizes and arrangements. The XML-format files we're more or less stuck with, since NVIDIA's flashing tools use them. I don't know how easy it would be to create a wic plugin to deal with the XML and the other oddities in the Jetson flashing process, but even if it isn't wic, some tool to assist with the XML wrangling would be great.

**Comment by windelbouwman** [comment]
It did not cross my mind that the flashing process requires interaction with the device. Sticking to the XML files seems like a good idea.

My original intention of this idea was to create a backend for WIC which will generate the proper XML files for flashing. In WIC it seems there is possibility to define partitions. Instead of creating a partitioning tool ourselves, it might be better to stick to WIC, as yocto users know it.

Another idea to ease partitioning might be to define variables and update the XML files according to those variables. For example, there would be these variables:

- A variable to enable dual boot A/B partitioning with a redundant rootFS partition
- A variable specifying the size of the root partition. This variable is already present.

**Reply by dwalkes** [reply]
Thanks for your comments @windelbouwman 

> Instead of creating a partitioning tool ourselves, it might be better to stick to WIC, as yocto users know it.

Completely agree, but the issue in this case is that the partitioning tools are already created by NVIDIA.  So we'd need a strategy to use our own separate tools but stay in sync with the changes they make to their tools and XML files generated as a part of L4T releases, which are at least partially closed source.  I like the idea of generating XML files.  I actually brought this up [a long time ago](https://github.com/OE4T/meta-tegra/pull/114#issuecomment-435678915) and @madisongh pointed out that the issue is less about being able to generate the content and more about being able to resynchronize when NVIDIA releases new versions with changes.

I think the best possible outcome here would be to convince NVIDIA to use tools like WIC as a part of their process, or at least formally support them.  To that end I've added to the [L4T Integration Issues](https://github.com/OE4T/meta-tegra/wiki/L4T-Integration-Issues) so we can at least track it as an idea.  

I'd be interested in any ideas you have to simplify the process ahead of this, as well.  Perhaps we could go from XML files in an NVIDIA L4T release to WIC variables, review changes and make any necessary modifications, and then use a WIC backend to dynamically generate XML files based on these variables.

**Reply by windelbouwman** [reply]
> I think the best possible outcome here would be to convince NVIDIA to use tools like WIC as a part of their process, or at least formally support them. To that end I've added to the [L4T Integration Issues](https://github.com/OE4T/meta-tegra/wiki/L4T-Integration-Issues) so we can at least track it as an idea.

This would be great, but I have low expectations here. This meta-tegra layer will be painful to maintain in any case.

**Comment by windelbouwman** [comment]
I was wading through the scripts / code to figure out where the partitioning was done. It appears the xml file from nvidia defines the layout. I think this is probably fine.

As mentioned elsewhere there are several ways to customize these flash files:
- keep a seperate copy with modifications (needs to be checked with each L4T release)
- apply a patch (needs to be checked with each L4T release)

The mender OTA approach keeps a full copy of the partition layout files for each board. To my opinion this should be moved into a configurable patch of the `tegra-bootfiles` recipe. This is the recipe which copies the xml layout files from the L4T release to the tegraflash folder. After making this copy, there is an opportunity to apply a patch to the xml file. This patch can then add the APP_b partition. The patching might be conditional on a variable or some setting. The benefit of this approach would be that it could be shared between rauc and mender usage, and is contained in the meta-tegra layer instead of other layers.

As mentioned this patch must be checked with each L4T release. Luckily the [tegra-bootfiles](https://github.com/OE4T/meta-tegra/blob/master/recipes-bsp/tegra-binaries/tegra-bootfiles_32.5.1.bb) is versioned to a specific L4T release, so version bumping results in a new recipe.

**Comment by windelbouwman** [comment]
Another idea / feature request I would like to propose is to move some code from `image_types_tegra.bbclass` to the `tegra-bootfiles` recipe. I have hard time to understand when exactly the image_types_tegra bbclass is invoked, things might become simpler when the xml bootfiles are prepared properly in the tegra bootfiles recipe.

**Reply by madisongh** [reply]
Sure, at least some of the `sed` substitutions could be done as part of the tegra-bootfiles.

**Comment by minicarpet** [comment]
Is there an integration of wic now ? I can see this thread is open since 2021

**Reply by dwalkes** [reply]
No wic integration. The biggest changes since this thread started is probably the default partition tables for redundant boot provided by nvidia and some additional capabilities in nvflashxmlparse.py to manipulate. I don't see wic integration playing nicely with NVIDIA's layout and I don't see them moving to this anytime soon.
